raiden/tests/unit/transfer/mediated_transfer/test_mediatorstate.py

```diff
@@ -2025,3 +2025,218 @@ def test_mediator_lock_expired_with_receive_lock_expired():
+    - M expires the lock.
+    - M should remove the lock from both secrethashes_to_lockedlocks
+      and secrethashes_to_unlocklocks.
+    """
+    amount = 10
+    block_number = 5
+    target = HOP2
+    expiration = 30
+    pseudo_random_generator = random.Random()
+
+    payer_channel = factories.make_channel(
+        partner_balance=amount,
+        partner_address=UNIT_TRANSFER_SENDER,
+        token_address=UNIT_TOKEN_ADDRESS,
+    )
+
+    payer_transfer = factories.make_signed_transfer_for(
+        payer_channel,
+        amount,
+        UNIT_TRANSFER_SENDER,
+        target,
+        expiration,
+        UNIT_SECRET,
+    )
+
+    channel1 = factories.make_channel(
+        our_balance=amount,
+        token_address=UNIT_TOKEN_ADDRESS,
+        partner_address=target,
+    )
+    channel_map = {
+        channel1.identifier: channel1,
+        payer_channel.identifier: payer_channel,
+    }
+    possible_routes = [factories.route_from_channel(channel1)]
+
+    init_mediator = ActionInitMediator(
+        possible_routes,
+        factories.route_from_channel(payer_channel),
+        payer_transfer,
+    )
+
+    iteration = mediator.state_transition(
+        None,
+        init_mediator,
+        channel_map,
+        pseudo_random_generator,
+        block_number,
+    )
+
+    transfer = payer_transfer
+    secrethash = transfer.lock.secrethash
+
+    block_lock_expired = expiration + DEFAULT_NUMBER_OF_BLOCK_CONFIRMATIONS
+
+    assert secrethash in payer_channel.partner_state.secrethashes_to_lockedlocks
+
+    # Reveal secret just before the lock expires
+    secret_reveal = ReceiveSecretReveal(UNIT_SECRET, target)
+
+    iteration = mediator.state_transition(
+        iteration.new_state,
+        secret_reveal,
+        channel_map,
+        pseudo_random_generator,
+        block_lock_expired,
+    )
+
+    # Make sure the lock was moved
+    assert secrethash not in payer_channel.partner_state.secrethashes_to_lockedlocks
+    assert secrethash in payer_channel.partner_state.secrethashes_to_unlockedlocks
+
+    block_expiration_number = transfer.lock.expiration + DEFAULT_NUMBER_OF_BLOCK_CONFIRMATIONS * 2
+    block = Block(
+        block_number=block_expiration_number,
+        gas_limit=1,
+        block_hash=factories.make_transaction_hash(),
+    )
+
+    iteration = mediator.state_transition(
+        iteration.new_state,
+        block,
+        channel_map,
+        pseudo_random_generator,
+        block_expiration_number,
+    )
+
+    # LockExpired should NOT remove the lock from unlockedlocks
+    # for a transfer that it already knows it already sent out the balance proof for.
+    # If we remove the lock this means that the mediator will expire the lock and
+    # lose funds due to the fact that the target already received the balance proof
+    # upon revealing the secret
+    assert secrethash in payer_channel.partner_state.secrethashes_to_unlockedlocks
```

Comment:
I would like to put an emphasis here on this line and comment above it to get opinions about this point.

The way i am thinking about it is:
- The mediator expires an unlocked/unclaimed lock if it receives a LockExpired message from the payer.
- The mediator does NOT expire unlocked/unclaimed locks if it detects them to be expired... so it only relies on the payer to declare the lock as expired, which means that the initiator is the only authority that can tell mediators to expire lock.

Reply:
I don't understand what you're saying here. A few notes:

- I wouldn't say that `LockedExpired` should or should not do anything, because it's just a message, I would talk in terms of the lock expiration block being reached.
- The mediator *must not* send a balance proof if the lock is in the danger zone, therefore the `iteration.events` above must **not** contain a `SendBalanceProof`
- The case where:
  - The mediator receives a secret reveal **before** the lock enters the danger zone, and the lock is unlocked
  - The payer hop does **not** send the off-chain unlock
  - The lock enters the danger zone and the secret registration is sent 
  - **and** the transaction is not mined in due time

  Is **the worst case scenario** but still possible, sadly there is nothing we can do at this point. My suggestion in the past was to add an event when this happens and continue with the protocol